cpu/sam0_common/periph/usbdev.c

```diff
@@ -0,0 +1,768 @@
+{
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+
+    if (ep->dir == USB_EP_DIR_IN) {
+        enable ? _enable_ep_stall_in(ep_reg)
+               : _disable_ep_stall_in(ep_reg);
+    }
+    else {
+        enable ? _enable_ep_stall_out(ep_reg)
+               : _disable_ep_stall_out(ep_reg);
+    }
+}
+
+usbopt_enable_t _ep_get_stall(usbdev_ep_t *ep)
+{
+    usbopt_enable_t res;
+    UsbDeviceEndpoint *ep_reg = _ep_reg_from_ep(ep);
+
+    if (ep->dir == USB_EP_DIR_IN) {
+        res = ep_reg->EPSTATUSSET.bit.STALLRQ1
+              ? USBOPT_ENABLE
+              : USBOPT_DISABLE;
+    }
+    else {
+        res = ep_reg->EPSTATUSSET.bit.STALLRQ0
+              ? USBOPT_ENABLE
+              : USBOPT_DISABLE;
+    }
+    return res;
+
+}
+
+static void _usbdev_ep_init(usbdev_ep_t *ep)
+{
+    _enable_ep_irq(ep);
+}
+
+static size_t _ep_get_available(usbdev_ep_t *ep)
+{
+    return _bank_from_ep(ep)->PCKSIZE.bit.BYTE_COUNT;
+}
+
+static int _usbdev_ep_get(usbdev_ep_t *ep, usbopt_ep_t opt,
+                          void *value, size_t max_len)
+{
+    (void)max_len;
+    int res = -ENOTSUP;
+    switch (opt) {
+        case USBOPT_EP_STALL:
+            assert(max_len == sizeof(usbopt_enable_t));
+            *(usbopt_enable_t *)value = _ep_get_stall(ep);
+            res = sizeof(usbopt_enable_t);
+            break;
+        case USBOPT_EP_AVAILABLE:
+            assert(max_len == sizeof(size_t));
+            *(size_t *)value = _ep_get_available(ep);
+            res = sizeof(size_t);
+            break;
+        default:
+            DEBUG("sam_usb: Unhandled get call: 0x%x\n", opt);
+            break;
+    }
+    return res;
+}
+
+static int _usbdev_ep_set(usbdev_ep_t *ep, usbopt_ep_t opt,
+                          const void *value, size_t value_len)
+{
+    (void)value_len;
+    int res = -ENOTSUP;
+    switch (opt) {
+        case USBOPT_EP_ENABLE:
+            assert(value_len == sizeof(usbopt_enable_t));
+            if (*((usbopt_enable_t *)value)) {
+                _ep_enable(ep);
+                _usbdev_ep_init(ep);
+            }
+            else {
+                _ep_disable(ep);
+            }
+            res = sizeof(usbopt_enable_t);
+            break;
+        case USBOPT_EP_STALL:
+            assert(value_len == sizeof(usbopt_enable_t));
+            _ep_set_stall(ep, *(usbopt_enable_t *)value);
+            res = sizeof(usbopt_enable_t);
+            break;
+        case USBOPT_EP_READY:
+            assert(value_len == sizeof(usbopt_enable_t));
+            if (*((usbopt_enable_t *)value)) {
+                _ep_unready(ep);
+            }
+            else {
+                _usbdev_ep_ready(ep, 0);
+            }
+            res = sizeof(usbopt_enable_t);
+            break;
+        default:
+            DEBUG("sam_usb: Unhandled set call: 0x%x\n", opt);
+            break;
+    }
+    return res;
+}
+
+static int _ep_unready(usbdev_ep_t *ep)
```

Comment:
nitpick: why `_ep_unready` and `_usbdev_ep_ready` ?

Reply:
Arbitrary, nothing specific, position of the moon etc.

Will fix to `_ep_ready`

Reply:
Never mind the above, I remember now, all the `static _usbdev_$SOMETHING` calls are used in the instantiated `usbdev_driver_t` struct